### PR TITLE
add uvc webcam support

### PR DESCRIPTION
**Pull request by cristiano-linvix**
Hello goodnight!

I made this library integration [AndroidUSBCamera](https://github.com/jiangdongguo/AndroidUSBCamera)

I hope it's useful for other people.

I would like to do the integration with the webcam's audio too, if I can I'll do another PR.

I'm not strong in java programming, I did what it takes to get it working.

https://github.com/react-native-webrtc/react-native-webrtc/issues/1282
https://github.com/react-native-webrtc/react-native-webrtc/issues/828

### CODE REVIEW

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    @Override
+    public void changeCaptureFormat(int i, int i1, int i2) {
+
+    }
+
+    @Override
+    public void dispose() {
+        Log.d(TAG, "UvcCapturer.dispose");
```
Comment (cristiano-linvix) [inline android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:95]: I need help to get to that point, to free the camera.

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+                if (!usbDevices.isEmpty()) {
+
+                    for (int i = 0; i < usbDevices.size(); i++) {
+                        UsbDevice dev = usbDevices.get(i);
+
+                        WritableMap params = Arguments.createMap();
+
+                        params.putString("facing", "uvc");
+                        params.putString("deviceId", "uvc-" + dev.getDeviceId());
```
Comment (cristiano-linvix) [inline android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:49]: I used the uvc prefix on the deviceId to be able to identify when looking for a uvc device in the CameraController.

Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:49]: There in UvcCapturer I remove the uvc prefix.

**Comment by 8BallBomBom** [comment]
Just a few thoughts.
After an initial look there are definitely some changes that could be made.

The `UvcCapturer` should probably extend the `AbstractVideoCaptureController` same as the `ScreenCaptureController` and `CameraCaptureController`, then it can also follow naming conventions like `USBCaptureController`.
You would also not then need the changes you've made within the `CameraCaptureController`.

But also rather than tying the grabbing logic in with `getUserMedia` instead make a function for it similar to `getDisplayMedia` which would also include some extra utility functions to handle creating the `USBCaptureController` and getting tracks.

Last but not least the actual exported `ReactMethod` can be handled within the `WebRTCModule` and we'd also need to work on some basic interfacing via typescript.

@saghul What do you think?

**Comment by cristiano-linvix** [comment]
Hello, thanks for the comment.

I think extending the AbstractVideoCaptureController really should be done.

Renaming the files to USBCapturer and USBEnumerator can be done if needed.

But, I don't see the need to create another function to handle the camera since it is used in getUserMedia, I think it "fits" the purpose of the function.

I await further information.

@saghul @8BallBomBom

**Comment by cristiano-linvix** [comment]
Hello, how are you there?

Did you look at this?

@saghul

**Comment by saghul** [comment]
Hi there! Thanks for sending the PR! I'm a bit busy now, I'll take a look as soon as I can.

**Comment by cristiano-linvix** [comment]
Hello, did you analyze this PR?

**Comment by saghul** [comment]
Not yet, sorry. It's been a bit busy. Hoping tkk on get to it next week though.

**Comment by saghul** [comment]
Can you please rebase so the new CI runs?

File: android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java [hunk 3]
```diff
@@ -115,19 +120,36 @@ public void onCameraSwitchError(String s) {
...
      * {@code deviceId} constraint
      */
     private VideoCapturer createVideoCapturer(String deviceId, String facingMode) {
         String[] deviceNames = cameraEnumerator.getDeviceNames();
         List<String> failedDevices = new ArrayList<>();
 
+        // If deviceId is specified, then it takes precedence over facingMode.
+        if (deviceId != null) {
+            // if has uvc into id, instance uvc capturer
+            if (deviceId.contains("uvc")) {
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java:139]: maybe we should have them start with `uvc:` instead? And then check if it _starts_ with that prefix.

File: android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java [hunk 4]
```diff
@@ -115,19 +120,36 @@ public void onCameraSwitchError(String s) {
...
+                String message = "Create UVC user-specified camera " + deviceId;
+                try {
+                    UvcCapturer videoCapturer = new UvcCapturer(deviceId, mUvcStrategy);
+                    Log.d(TAG, message + " succeeded");
+                    return videoCapturer;
+                } catch (Throwable tr) {
+                    failedDevices.add(deviceId);
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java:146]: The fallback code won't work with this I think, because we are not enumerating the UVC cameras.

File: android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java [hunk 5]
```diff
@@ -201,7 +216,9 @@ void getUserMedia(
 
             CameraCaptureController cameraCaptureController = new CameraCaptureController(
                 cameraEnumerator,
-                videoConstraintsMap);
+                videoConstraintsMap,
+                cameraUvcStrategy
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java:220]: can't we pass the enumerator instead? feels more natural to have 1 "meta" enumerator.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    private String deviceId;
+
+    private SurfaceTextureHelper surfaceTextureHelper;
+    private CapturerObserver capturerObserver;
+
+    private int UVC_PREVIEW_WIDTH = 1024;
+    private int UVC_PREVIEW_HEIGHT = 768;
+    private int UVC_PREVIEW_FPS = 30;
+
+    CameraUvcStrategy mUvcStrategy; // From Jiang Dongguo's AUSBC library
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:34]: Drop the commend, it's not helpful.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+    private static final String TAG = WebRTCModule.TAG;
+
+    private String deviceId;
+
+    private SurfaceTextureHelper surfaceTextureHelper;
+    private CapturerObserver capturerObserver;
+
+    private int UVC_PREVIEW_WIDTH = 1024;
+    private int UVC_PREVIEW_HEIGHT = 768;
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:31]: This is a very weird default resolution. Shall we make it 1280x720 aka 720p ?

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    private int UVC_PREVIEW_FPS = 30;
+
+    CameraUvcStrategy mUvcStrategy; // From Jiang Dongguo's AUSBC library
+
+    public UvcCapturer(String deviceId, CameraUvcStrategy uvcStrategy) {
+        this.deviceId = deviceId.replaceAll("\\D", "");
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:37]: Why is this necessary?

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+        Log.d(TAG, "UvcCapturer.UvcCapturer->DeviceID " + this.deviceId);
+
+        try {
+            mUvcStrategy = uvcStrategy;
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:42]: How could this fail? It's just a variable assignment.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+    @Override
+    public void startCapture(int width, int height, int fps) {
+        if (mUvcStrategy != null) {
+            // stop before to free last usb instance
+            if (mUvcStrategy.isCameraOpened()) {
+                mUvcStrategy.stopPreview();
+            }
+
+            UVC_PREVIEW_WIDTH = width;
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:62]: why do we need to save these here?

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+            UVC_PREVIEW_HEIGHT = height;
+            UVC_PREVIEW_FPS = fps;
+
+            mUvcStrategy.addPreviewDataCallBack(new IPreviewDataCallBack() {
+                @Override
+                public void onPreviewData(@Nullable byte[] bytes, @NonNull DataFormat dataFormat) {
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:68]: Why is dataFormat not being used?

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java (hunk 1 above)
Comment (saghul) [reply android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:95]: What do you mean?

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+    private CameraRequest getCameraRequest() {
+        return new CameraRequest
+                .Builder()
+                .setCameraId(this.deviceId)
+                // .setFrontCamera(false)
+                .setContinuousAFModel(true)
+                .setContinuousAutoModel(true)
+                .setPreviewWidth(UVC_PREVIEW_WIDTH)
+                .setPreviewHeight(UVC_PREVIEW_HEIGHT)
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:112]: You are not specifying the fps here. Is there a way to do that?

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+import org.webrtc.SurfaceViewRenderer;
+
+import java.util.List;
+
+public class UvcEnumerator {
+
+    private static final String TAG = WebRTCModule.TAG;
+
+    CameraUvcStrategy mUvcStrategy; // From Jiang Dongguo's AUSBC library
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:23]: Drop the comment please.

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+    public UvcEnumerator(CameraUvcStrategy cameraUvcStrategy) {
+        try {
+            mUvcStrategy = cameraUvcStrategy;
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:27]: How can this ever fail?

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+            if (usbDevices != null) {
+
+                if (!usbDevices.isEmpty()) {
+
+                    for (int i = 0; i < usbDevices.size(); i++) {
+                        UsbDevice dev = usbDevices.get(i);
+
+                        WritableMap params = Arguments.createMap();
+
+                        params.putString("facing", "uvc");
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:48]: Can't we get the right facing mode? If not, let's just say "environment".

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java (hunk 2 above)
Comment (saghul) [reply android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:49]: Good idea!

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+                        params.putString("label", dev.getProductName() != null ? dev.getProductName() : dev.getDeviceName());
+                        params.putString("kind", "videoinput");
+
+                        Log.d(TAG, "UvcEnumerator.enumerateDevices: " + params);
+
+                        array.pushMap(params);
+                    }
+
+                } else {
+                    Log.d(TAG, "UvcEnumerator.enumerateDevices: usbDevices is empty");
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:60]: I'd remove this log line.

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+                }
+
+            } else {
+                Log.d(TAG, "UvcEnumerator.enumerateDevices: usbDevices is null");
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:64]: Ditto.

**Review by saghul — CHANGES_REQUESTED** [review]
Finally managed to do a full review, please take a look!

**Comment by saghul** [comment]
Question: is there a chance a camera is detected by both enumerators?

**Comment by cristiano-linvix** [comment]
Hello @saghul, I'll start on the changes as soon as possible, I'm a little busy this month.

Thank you!

File: android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java (hunk 3 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java:139]: It would be to change the uvc- for uvc: ?

File: android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java (hunk 4 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java:146]: Okay, so I'll remove the snippet.

File: android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java (hunk 5 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java:220]: Sorry, that I didn't understand.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java (hunk 6 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:34]: I will remove the comment.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java (hunk 7 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:31]: WebCam resolutions are very different between models, I'll try using 720 instead of 728 and see if it works.

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java (hunk 14 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:27]: I'll look into it.

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java (hunk 15 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:48]: I'll switch to environment.

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java (hunk 16 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:60]: I will.

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java (hunk 17 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:64]: I will.

**Comment by saghul** [comment]
No worries, thank you!

**Comment by cristiano-linvix** [comment]
> Question: is there a chance a camera is detected by both enumerators?

It depends on the android device, if it supports UVC devices cameras, it may appear in the first enumerator too.

https://source.android.com/docs/core/camera/external-usb-cameras#implementation

**Comment by saghul** [comment]
I feared that. In that case, we'll need to filter it, and prefer the non-UVC one, I'd say.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java (hunk 11 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:62]: The data pattern that is received is always the same, you don't need to detect it.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java (hunk 12 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:106]: Maximum FPS is 30.
Its not possible to change. :(

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java (hunk 10 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:56]: The getCameraRequest function uses this variable to get the webcam that uses this resolution.
The startCapture function receives the size there from the controller.

**Comment by cristiano-linvix** [comment]
@saghul I sent it for your analysis again, as soon as you can look, thank you.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+public class UvcCapturer implements VideoCapturer {
+
+    private static final String TAG = WebRTCModule.TAG;
+
+    private String deviceId;
+
+    private SurfaceTextureHelper surfaceTextureHelper;
+    private CapturerObserver capturerObserver;
+
+    private int UVC_PREVIEW_WIDTH = 1280;
```
Comment (cristiano-linvix) [inline android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:30]: UVC Capturer resolutions are different, it looks like they are screen resolutions.

Logcat
```log
getAllPreviewSizes = [Size(1920x1080@ 0.0,type:4,frame:0,index:0,[]), Size(1024x768@ 0.0,type:4,frame:0,index:1,[]), Size(1024x576@ 0.0,type:4,frame:0,index:2,[]), Size(960x720@ 0.0,type:4,frame:0,index:3,[]), Size(864x480@ 0.0,type:4,frame:0,index:4,[]), Size(800x600@ 0.0,type:4,frame:0,index:5,[]), Size(640x480@ 0.0,type:4,frame:0,index:6,[]), Size(320x240@ 0.0,type:4,frame:0,index:7,[]), Size(1920x1080@ 0.0,type:4,frame:0,index:8,[])]
```

Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:30]: It really needs to be 1024x768.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java (hunk 7 above)
Comment (cristiano-linvix) [reply android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:31]: It really needs to be 1024x768.

The webcam resolutions are different, like monitor resolution.

logcat
```log
getAllPreviewSizes = [Size(1920x1080@ 0.0,type:4,frame:0,index:0,[]), Size(1024x768@ 0.0,type:4,frame:0,index:1,[]), Size(1024x576@ 0.0,type:4,frame:0,index:2,[]), Size(960x720@ 0.0,type:4,frame:0,index:3,[]), Size(864x480@ 0.0,type:4,frame:0,index:4,[]), Size(800x600@ 0.0,type:4,frame:0,index:5,[]), Size(640x480@ 0.0,type:4,frame:0,index:6,[]), Size(320x240@ 0.0,type:4,frame:0,index:7,[]), Size(1920x1080@ 0.0,type:4,frame:0,index:8,[])]
```

**Comment by cristiano-linvix** [comment]
> I feared that. In that case, we'll need to filter it, and prefer the non-UVC one, I'd say.

The function is:
```java
public bool supportsExternalCameras() {
        return this.context.getPackageManager().hasSystemFeature(PackageManager.FEATURE_CAMERA_EXTERNAL);        
}
```

Put inside the Camera Capture Controller?
I would have to pass the context into the class.

**Comment by saghul** [comment]
> The function is:
> 
> ```java
> public bool supportsExternalCameras() {
>         return this.context.getPackageManager().hasSystemFeature(PackageManager.FEATURE_CAMERA_EXTERNAL);        
> }
> ```
> 
> Put inside the Camera Capture Controller? I would have to pass the context into the class.

But that would only tell you if external cameras are supported, right? The problem is if they are, and both enumerators list them. How do you plan on filtering them?

File: android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java [hunk 19]
```diff
@@ -31,14 +34,16 @@ public class CameraCaptureController extends AbstractVideoCaptureController {
      */
     private final CameraEventsHandler cameraEventsHandler = new CameraEventsHandler();
 
-    public CameraCaptureController(CameraEnumerator cameraEnumerator, ReadableMap constraints) {
+    public CameraCaptureController(CameraEnumerator cameraEnumerator, ReadableMap constraints, CameraUvcStrategy uvcStrategy) {
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java:37]: Can you please make the uvcStrategy the 2nd parameter?

File: android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java [hunk 20]
```diff
@@ -86,7 +91,7 @@ protected VideoCapturer createVideoCapturer() {
      * Helper function which tries to switch cameras until the desired facing mode is found.
      *
      * @param desiredFrontFacing - The desired front facing value.
-     * @param tries - How many times to try switching.
+     * @param tries              - How many times to try switching.
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java:94]: Please avoid making any unrelated changes. It complicates the diff for no good reason. I'm working on integrating clang-format to take care of this automatically.

File: android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java [hunk 21]
```diff
@@ -115,19 +120,35 @@ public void onCameraSwitchError(String s) {
...
      * {@code deviceId} constraint
      */
     private VideoCapturer createVideoCapturer(String deviceId, String facingMode) {
         String[] deviceNames = cameraEnumerator.getDeviceNames();
         List<String> failedDevices = new ArrayList<>();
 
+        // If deviceId is specified, then it takes precedence over facingMode.
+        if (deviceId != null) {
+            // if has uvc into id, instance uvc capturer
+            if (deviceId.contains("uvc:")) {
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/CameraCaptureController.java:139]: Let's be precise and check if it starts with the prefix, not if it contains it.

File: android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java (hunk 5 above)
Comment (saghul) [reply android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java:220]: I think we can create the strategy inside the enumerator, and provide a getter for it.

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+
+    ReadableArray enumerateDevices() {
+        WritableArray array = Arguments.createArray();
+
+        try {
+            List<UsbDevice> usbDevices = mUvcStrategy.getUsbDeviceList(null);
+
+            if (usbDevices != null) {
+
+                if (!usbDevices.isEmpty()) {
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:37]: Is this necessary? The for look won't run if it's empty and that's that...

File: android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+                        Log.d(TAG, "UvcEnumerator.enumerateDevices: " + params);
+
+                        array.pushMap(params);
+                    }
+
+                }
+
+            }
+
+        } catch (Throwable tr) {
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/UvcEnumerator.java:59]: Can you please try to make try-catch block as narrow was possible? That is, please only protect the calls that can throw, not the whole thing.

File: android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java [hunk 24]
```diff
@@ -138,8 +145,16 @@ private void checkMandatoryConstraints(MediaConstraints peerConstraints) {
 
     ReadableArray enumerateDevices() {
         WritableArray array = Arguments.createArray();
+
+        ReadableArray uvcDevices = uvcEnumerator.enumerateDevices();
         String[] devices = cameraEnumerator.getDeviceNames();
 
+        if (uvcDevices.size() > 0) {
+            for (int j = 0; j < uvcDevices.size(); j++) {
+                array.pushMap(uvcDevices.getMap(j));
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java:154]: We need to filter the devices. maybe you can do something like `uvcEnumerator.enumerateDevices(nativeDevices);` where `nativeDevices` is `cameraEnumerator.getDeviceNames();` and you can do the filtering there, if they match.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java (hunk 7 above)
Comment (saghul) [reply android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:31]: Alternatively, figure it out dynamically by querying the supported sizes and picking the one closest to the request resolution that is passed in startCapture, that would make it even better.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java (hunk 10 above)
Comment (saghul) [reply android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:56]: Please pass height, width, fps as parameters to `getCameraRequest()` and get rid off these.

File: android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java (hunk 1 above)
Comment (saghul) [reply android/src/main/java/com/oney/WebRTCModule/UvcCapturer.java:89]: Doesn't this dispose method work? Alternatively, we can do it in start / stop. Start re-creates what's necessary and stop destroys it.

**Review by saghul — CHANGES_REQUESTED** [review]
Left some more comments, this is looking better!

**Comment by saghul** [comment]
The PR needs rebasing and addressing some of my comments still.

**Comment by cristiano-linvix** [comment]
Hello @saghul, sorry for the delay, I ended up having to take time off work because of a medical procedure, but I'll be back in April.

**Comment by saghul** [comment]
No worries, no rush at all.

Hope it all goes well!

**Comment by cristiano-linvix** [comment]
Hello everybody!

I'm going to close this pull request, it's been open for a long time, and we need to implement this in the latest version of the project.

When I have time again, I start the project again and make a new pull request.

Thank you all!
